internal/webhooks/etcdcluster_webhook.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2024 The etcd-operator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhooks
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/aenix-io/etcd-operator/api/v1alpha1"
+	"github.com/aenix-io/etcd-operator/internal/controller/factory"
+)
+
+// log is for logging in this package.
+var etcdclusterlog = logf.Log.WithName("etcdcluster-resource")
+
+type EtcdCluster struct{}
```

Comment:
What is the purpose of moving webhooks to `internal` package? I suppose webhooks depend on CRD structure and should stay in `v1alpha1` for now

Reply:
@sircthulhu I had the same question. As far as I understand, webhook imports both packages: api and factory. Hence, since factory imports api, in case webhooks would be a part of api package there woulb be circular dependencies.
Which definitely can be solved be re-organising of new feature.
I'm with you on the point to keep initial structure as is.

Reply:
Yep, to avoid circular imports

Reply:
Also this approach allows add some interfaces into 
```type EtcdCluster struct{} ```
during initialization.

Reply:
I can move webhooks files back. Then make isolation based on interface.
It will look like this:
```
// +k8s:deepcopy-gen=false
type FactoryValidator interface {
	ValidatePodExtraArgs(*EtcdCluster) error
}

// +k8s:deepcopy-gen=false
type EtcdClusterHook struct {
	FactoryValidator
}

// SetupWebhookWithManager will setup the manager to manage the webhooks
func (r *EtcdClusterHook) SetupWebhookWithManager(mgr ctrl.Manager) error {
	return ctrl.NewWebhookManagedBy(mgr).
		For(&EtcdCluster{}).
		WithValidator(r).
		WithDefaulter(r).
		Complete()
}
```
Then  I need to make validation function  as a method.
```
type Validator struct{}

func (v Validator) ValidatePodExtraArgs(cluster *etcdaenixiov1alpha1.EtcdCluster) error
```

But, I don't like how it looks like...

Reply:
Please help me to take a choice.

Reply:
I would put `ValidatePodExtraArgs` to `etcdcluster_webhook.go` as it just performs validation, not logic for generating statefulset

Reply:
The knowledge about how to build arguments belongs to business logic.
I want to store base arguments only in one packet to avoid business logic splitting. (DRY)
It seems that validation process belongs to the factory packet too.

There are three ways:
1. Repeat two absolutely the same sets of arguments in different packages. 
2. Move webhooks into particular package and use custom hooks  (as I did)
3. Use custom hooks isolated on interfaces. It needs to transform factory package to sets of methods.

Reply:
> The knowledge about how to build arguments belongs to business logic.

I agree. But validation rules should belong to webhooks in CRD's package.

In `v1alpha1` we validate keys: reject prefix dashes and reject intersection with flags used by the operator. In `factory` we use keys and values from map to build flags for statefulset.

Can't find any logic duplications. Am I missing something?

Reply:
I like Idea to separate webhooks from API. Then we can setup API as separate go module that contains only definitions. If later someone want to include this module, he will depend only on API dependencies. Not all operator dependencies

Reply:
> In `v1alpha1` we validate keys: reject prefix dashes and reject intersection with flags used by the operator. In `factory` we use keys and values from map to build flags for statefulset.

That case implies store the flags  in two places.

Reply:
@sergeyshevch I don't like moving webhooks to internal as for each api version there will be different webhooks. If we use interfaces and additional go module in api, users will pull whole project with it's dependencies. Do we really need this validation for extra args at all?

Operator must override flags and ignore if conflict happens


@andreykont 
> That case implies store the flags in two places.

Actually, we do not store this flags anywhere, we just use them in one place and say they're prohibited to use in another place. You can't do it any other way as validation is separate from usage even in one package. Checking validation after implementing changes to business logic is really common I think and it won't lead to problems. If it will, it's user's fault :)

Reply:
https://github.com/aenix-io/etcd-operator/pull/99#issuecomment-2032241368